### PR TITLE
raft: don't step down to newer term while fortifying

### PR DESCRIPTION
**Pull request by nvb**
Informs #132762.

This commit fixes a bug where a peer would step down to a newer term while fortifying a leader. This would break the leader fortification promise, which is critical for the leader lease disjointness property.

Instead, we now ignore the message as a follower and let the leader handle the process of stepping down, safely defortifying, and eventually stepping back up leader in the new term.

A subsequent commit will combine this logic with the other leader lease check for votes and pre-votes.

Release note: None

### CODE REVIEW

**Review by iskettaneh — COMMENTED** [review]
> Reviewed 1 of 1 files at r1, 3 of 3 files at r2, 30 of 30 files at r3, 21 of 23 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134922)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/raft/testdata/leader_step_down_stranded_peer_prevote.txt` line 140 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/134922#-OBVrHIR3VoPDV9m_JCd:-OBVrHIR3VoPDV9m_JCe:b-a7bvqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/06ccd1161893d8d18108e2310d0e1e53f523b7d5/pkg/raft/testdata/leader_step_down_stranded_peer_prevote.txt#L140)):*
> ```help
> INFO 3 [logterm: 1, index: 10] sent MsgPreVote request to 2 at term 2
> 
> stabilize 3
> ```

nit: can we stabilize both 2 and 3 at the same time?
___
*[`pkg/raft/testdata/leader_step_down_stranded_peer.txt` line 358 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/134922#-OBVqcRODz7TEWnRdzyH:-OBVqcRODz7TEWnRdzyI:b-obtj1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/06ccd1161893d8d18108e2310d0e1e53f523b7d5/pkg/raft/testdata/leader_step_down_stranded_peer.txt#L358)):*
> ```help
>   3->1 MsgAppResp Term:4 Log:1/11 Rejected (Hint: 10) Commit:10
>   DEBUG 1 received MsgAppResp(rejected, hint: (index 10, term 1)) from 3 for index 11
>   3->1 MsgAppResp Term:4 Log:0/12 Commit:12
> ```

nit: maybe print raft-state at the end of the test and the prevote test
___
*[`pkg/raft/raft.go` line 1562 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/134922#-OBVu2gH-4xzIHkWZLcB:-OBVu2gH-4xzIHkWZLcC:b-7qvx9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/61226e53cdeaacd473bc6f3c6506711669d862a1/pkg/raft/raft.go#L1562)):*
> ```Go
> 				lead = m.From
> 			}
> 			r.logMsgHigherTerm(m, "new leader indicated, advancing term")
> ```

Should this log line be before the deFortify() call? This was when reading the log, we can more easily see what caused the defortification to happen.

**Review by nvb — COMMENTED** [review]
With this change, we now reliably pass `kv/splits/nodes=3/quiesce=false/lease=leader`, added in https://github.com/cockroachdb/cockroach/pull/133216.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134922#-:-OBW4hxw1Uh7CT0h2oTV:bvvhf11)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @iskettaneh)

___
*[`pkg/raft/raft.go` line 1562 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/134922#-OBVu2gH-4xzIHkWZLcB:-OBW4Pz1A2TfkmBAgLtm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/61226e53cdeaacd473bc6f3c6506711669d862a1/pkg/raft/raft.go#L1562)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

Should this log line be before the deFortify() call? This was when reading the log, we can more easily see what caused the defortification to happen.
</blockquote></details>

Done.
___
*[`pkg/raft/testdata/leader_step_down_stranded_peer.txt` line 358 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/134922#-OBVqcRODz7TEWnRdzyH:-OBW5PcP5qNa5kglAZ4D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/06ccd1161893d8d18108e2310d0e1e53f523b7d5/pkg/raft/testdata/leader_step_down_stranded_peer.txt#L358)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

nit: maybe print raft-state at the end of the test and the prevote test
</blockquote></details>

Done.
___
*[`pkg/raft/testdata/leader_step_down_stranded_peer_prevote.txt` line 140 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/134922#-OBVrHIR3VoPDV9m_JCd:-OBW4UVM2cBfQQmBOkZg:bmwaph9) ([raw file](https://github.com/cockroachdb/cockroach/blob/06ccd1161893d8d18108e2310d0e1e53f523b7d5/pkg/raft/testdata/leader_step_down_stranded_peer_prevote.txt#L140)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

nit: can we stabilize both 2 and 3 at the same time?
</blockquote></details>

No, if we stabilized both at the same time then 3's MsgVotes would reach 2 and it would win the election.

**Review by iskettaneh — APPROVED** [review]
> Reviewed 34 of 34 files at r6, 3 of 3 files at r7, 30 of 30 files at r8, 23 of 23 files at r9, 4 of 4 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134922)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

**Comment by nvanbenschoten** [comment]
TFTR!

bors r+

File: pkg/raft/raft.go [hunk 1]
```diff
@@ -1742,6 +1719,55 @@ func (r *raft) Step(m pb.Message) error {
...
+		(m.Type == pb.MsgPreVote || m.Type == pb.MsgVote)
+
+	inFortifyLease := r.supportingFortifiedLeader() &&
+		// NB: If the peer that's campaigning has an entry in its log with a
+		// higher term than what we're aware of, then this conclusively proves
+		// that a new leader was elected at a higher term. We never heard from
+		// this new leader (otherwise we'd have bumped r.Term in response).
+		// However, any fortification we're providing to a leader that has been
+		// since dethroned is pointless.
+		m.LogTerm <= r.Term
```
Comment (pav-kv) [inline pkg/raft/raft.go:1750]: Should we check the message type before interpreting `LogTerm`?

**Pull request merged**